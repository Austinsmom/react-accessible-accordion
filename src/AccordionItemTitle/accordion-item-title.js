// @flow

import React, { Component } from 'react';
import type { Node } from 'react';
import classNames from 'classnames';
import { inject, observer } from 'mobx-react';

type AccordionItemTitleProps = {
    children: Node,
    className: string,
    hideBodyClassName: string,
    accordionStore: {
        items: Array<Object>,
        accordion: boolean,
        onChange: Function,
        setExpanded: (string | number, boolean) => void,
    },
    itemkey: string | number,
};

type AccordionItemTitleState = {};

export class AccordionItemTitle extends Component<
    AccordionItemTitleProps,
    AccordionItemTitleState,
> {
    static accordionElementName = 'AccordionItemTitle';

    static defaultProps = {
        className: 'accordion__title',
        hideBodyClassName: '',
    };

    handleClick = () => {
        const { itemkey, accordionStore } = this.props;
        const { accordion, onChange, items } = accordionStore;
        const foundItem = items.find(item => item.itemkey === itemkey);
        if (!foundItem) return;

        if (accordion) {
            const newValue = Boolean(!foundItem.expanded);
            items.forEach(item => {
                this.props.accordionStore.setExpanded(item.itemkey, false);
                if (item.itemkey === foundItem.itemkey) {
                    this.props.accordionStore.setExpanded(
                        foundItem.itemkey,
                        newValue,
                    );
                }
            });

            onChange(foundItem.itemkey);
        } else {
            this.props.accordionStore.setExpanded(
                foundItem.itemkey,
                !foundItem.expanded,
            );
            onChange(
                this.props.accordionStore.items
                    .filter(item => item.expanded)
                    .map(item => item.itemkey),
            );
        }
    };

    handleKeyPress = (evt: SyntheticInputEvent<HTMLButtonElement>) => {
        if (evt.charCode === 13 || evt.charCode === 32) {
            this.handleClick();
        }
    };

    render() {
        const { items, accordion } = this.props.accordionStore;
        const { itemkey, children, className, hideBodyClassName } = this.props;
        const foundItem = items.find(item => item.itemkey === itemkey);
        if (!foundItem) return null;

        const { itemUuid, expanded } = foundItem;

        const id = `accordion__title-${itemUuid}`;
        const ariaControls = `accordion__body-${itemUuid}`;
        const role = accordion ? 'tab' : 'button';
        const titleClassName = classNames(className, {
            [hideBodyClassName]: hideBodyClassName && !expanded,
        });

        if (role === 'tab') {
            return (
                <div
                    id={id}
                    aria-selected={expanded}
                    aria-controls={ariaControls}
                    className={titleClassName}
                    onClick={this.handleClick}
                    role={role}
                    tabIndex="0" // eslint-disable-line jsx-a11y/no-noninteractive-tabindex
                    onKeyPress={this.handleKeyPress}
                >
                    {children}
                </div>
            );
        }
        return (
            <div
                id={id}
                aria-expanded={expanded}
                aria-controls={ariaControls}
                className={titleClassName}
                onClick={this.handleClick}
                role={role}
                tabIndex="0" // eslint-disable-line jsx-a11y/no-noninteractive-tabindex
                onKeyPress={this.handleKeyPress}
            >
                {children}
            </div>
        );
    }
}

export default inject('accordionStore', 'itemkey')(
    observer(AccordionItemTitle),
);
