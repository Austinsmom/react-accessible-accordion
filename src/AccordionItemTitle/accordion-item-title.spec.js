// @flow

import React from 'react';
import renderer from 'react-test-renderer';
// import { mount } from 'enzyme';
import { Provider } from 'unstated';
import AccordionItemTitle from './accordion-item-title';
import ItemContainer from '../ItemContainer/ItemContainer';
import AccordionContainer from '../AccordionContainer/AccordionContainer';

describe('AccordionItemTitle', () => {
    let accordionStore;
    let itemStore;
    // let onChange;

    beforeEach(() => {
        // onChange = jest.fn();

        const itemContainer = new ItemContainer();
        itemContainer.resetNextUuid();
        const accordionContainer = new AccordionContainer();
        accordionContainer.setAccordion(false);
        accordionContainer.setOnChange(jest.fn());
        accordionStore = accordionContainer;
        itemStore = itemContainer;

        accordionStore.addItem({
            uuid: 'item-one-uuid',
            expanded: false,
            disabled: false,
        });
        accordionStore.addItem({
            uuid: 'item-two-uuid',
            expanded: true,
            disabled: false,
        });
    });

    it('renders correctly with min params', () => {
        const tree = renderer
            .create(
                <Provider inject={[accordionStore, itemStore]}>
                    <AccordionItemTitle>
                        <div>Fake Title</div>
                    </AccordionItemTitle>
                </Provider>,
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    // it('renders correctly with different className', () => {
    //     const className = 'className';
    //     const wrapper = mount(
    //         <AccordionItemTitle
    //             className={className}
    //             accordionStore={accordionStore}
    //             uuid="item-one-uuid"
    //         />,
    //     );
    //     expect(wrapper.find('div').hasClass(className)).toEqual(true);
    // });

    // it('renders with different hideBodyClassName', () => {
    //     const hideBodyClassName = 'hideBodyClassName';
    //     const wrapper = mount(
    //         <AccordionItemTitle
    //             hideBodyClassName={hideBodyClassName}
    //             accordionStore={accordionStore}
    //             uuid="item-one-uuid"
    //         />,
    //     );
    //     expect(wrapper.find('div').hasClass(hideBodyClassName)).toEqual(true);
    // });

    // it("doesn't present hideBodyClassName when collapsed", () => {
    //     const hideBodyClassName = 'hideBodyClassName';
    //     const wrapper = mount(
    //         <AccordionItemTitle
    //             hideBodyClassName={hideBodyClassName}
    //             accordionStore={accordionStore}
    //             uuid="item-two-uuid"
    //         />,
    //     );
    //     expect(wrapper.find('div').hasClass(hideBodyClassName)).toEqual(false);
    // });

    // it('renders correctly when pressing enter', async () => {
    //     const wrapper = mount(
    //         <AccordionItemTitle
    //             accordionStore={accordionStore}
    //             uuid="item-one-uuid"
    //         >
    //             Fake Title
    //         </AccordionItemTitle>,
    //     );

    //     wrapper.find('div').simulate('keypress', { charCode: 13 });

    //     expect(onChange.mock.calls.length).toEqual(1);
    //     expect(
    //         accordionStore.items.filter(item => item.expanded === true).length,
    //     ).toEqual(2);
    // });

    // it('renders correctly when pressing space', async () => {
    //     const wrapper = mount(
    //         <AccordionItemTitle
    //             accordionStore={accordionStore}
    //             uuid="item-one-uuid"
    //         >
    //             Fake Title
    //         </AccordionItemTitle>,
    //     );

    //     wrapper.find('div').simulate('keypress', { charCode: 32 });

    //     expect(onChange.mock.calls.length).toEqual(1);
    //     expect(
    //         accordionStore.items.filter(item => item.expanded === true).length,
    //     ).toEqual(2);
    // });

    // it('renders correctly when pressing another key', async () => {
    //     const wrapper = mount(
    //         <AccordionItemTitle
    //             accordionStore={accordionStore}
    //             uuid="item-one-uuid"
    //         >
    //             Fake Title
    //         </AccordionItemTitle>,
    //     );

    //     wrapper.find('div').simulate('keypress', { charCode: 35 });

    //     expect(onChange.mock.calls.length).toEqual(0);
    //     expect(
    //         accordionStore.items.filter(item => item.expanded === true).length,
    //     ).toEqual(1);
    // });

    // it('renders null if an associated AccordionItem is not registered in accordionStore', () => {
    //     const className = 'className';
    //     const wrapper = mount(
    //         <Provider accordionStore={accordionStore} uuid="foo">
    //             <AccordionItemTitle className={className}>
    //                 <div>Fake body</div>
    //             </AccordionItemTitle>
    //         </Provider>,
    //     );

    //     expect(
    //         wrapper.findWhere(item => item.className === className).length,
    //     ).toEqual(0);
    // });

    // it('respects arbitrary user-defined props', () => {
    //     const wrapper = mount(
    //         <AccordionItemTitle
    //             accordionStore={accordionStore}
    //             uuid="item-one-uuid"
    //             lang="en"
    //         >
    //             Fake Title
    //         </AccordionItemTitle>,
    //     );

    //     expect(wrapper.find('div').instance().lang).toEqual('en');
    // });
});
