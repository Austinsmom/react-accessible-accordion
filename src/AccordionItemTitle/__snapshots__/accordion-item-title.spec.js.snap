// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionItemTitle renders correctly with min params 1`] = `
<div
  aria-controls="accordion__body-item-one-uuid"
  aria-expanded={false}
  className="accordion__title"
  disabled={false}
  id="accordion__title-item-one-uuid"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="button"
  tabIndex="0"
>
  <div>
    Fake Title
  </div>
</div>
`;
